### PR TITLE
Parallelize Repository Cleanup Actions

### PR DESCRIPTION
**Pull request by original-brownbear**
Deleting root blobs and unreferenced indices can safely happen in parallel,
no need to have both operations run sequentially when they preclude all other
repository operations.
Also, this PR dries up the cleanup action invocation between the cleanup and delete logic and adds a little more documentation.

Preliminary for #46250 (as we'll have three delete paths there, so extracting the cleanup into a method was needed imo).

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 1]
```diff
@@ -418,16 +419,38 @@ public void deleteSnapshot(SnapshotId snapshotId, long repositoryStateId, Action
...
+     * parameters {@code foundIndices}, {@code rootBlobs}
+     *
+     * @param foundIndices all indices blob containers found in the repository before {@code newRepoData} was written
+     * @param rootBlobs    all blobs found directly under the repository root
+     * @param newRepoData  new repository data that was just written
+     * @param listener     listener to invoke with the combined {@link DeleteResult} of all blobs removed in this operation
+     */
+    private void cleanupStaleBlobs(Map<String, BlobContainer> foundIndices, Map<String, BlobMetaData> rootBlobs,
+                                   RepositoryData newRepoData, ActionListener<DeleteResult> listener) {
+        final StepListener<DeleteResult> deleteIndicesListener = new StepListener<>();
```
[the remark this answers is not in the document]
Comment (tlrx) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Thanks, I find it easier to read now.

**Review by tlrx — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/1

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/1

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/1

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/1

**Comment by original-brownbear** [comment]
Thanks Tanguy!

**Pull request merged**